src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -381,6 +369,103 @@ func (a *StorageMinerActorCode_I) ProcessVerifiedElectionPoSt(rt Runtime) InvocO
 	return a._onSuccessfulPoSt(rt)
 }
 
+// called by verifier to update miner state on successful surprise post
+// after it has been verified in the storage_mining_subsystem
+func (a *StorageMinerActorCode_I) ProcessSurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) InvocOutput {
+	TODO() // TODO: validate caller
+
+	if !a._verifySurprisePoSt(rt, onChainInfo) {
+		rt.Abort(
+			exitcode.UserDefinedError(exitcode.PoStVerificationFailed),
+			"PoSt verification failed")
+	}
+
+	// Ensure pledge collateral satisfied
+	// otherwise, abort ProcessVerifiedSurprisePoSt
+	a._ensurePledgeCollateralSatisfied(rt)
+
+	return a._onSuccessfulPoSt(rt)
+
+}
+
+func (a *StorageMinerActorCode_I) GetWorkerKeyByMinerAddress(minerAddr addr.Address) filcrypto.VRFPublicKey {
+	panic("TODO")
+}
+
+func (a *StorageMinerActorCode_I) _verifySurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) bool {
+	h, st := a.State(rt)
+	info := st.Info()
+
+	// 1. Check that the miner in question is currently being challenged
+	if !st._isChallenged() {
+		// TODO: determine proper error here and error-handling machinery
+		// rt.Abort("cannot SubmitSurprisePoSt when not challenged")
+		rt.AbortStateMsg("Invalid Surprise PoSt. Miner not challenged.")
+	}
+
+	// 2. Check that the challenge has not expired
+	// Check that miner can still submit (i.e. that the challenge window has not passed)
+	// This will prevent miner from submitting a Surprise PoSt past the challenge period
+	if st._challengeHasExpired(rt.CurrEpoch()) {
+		rt.AbortStateMsg("Invalid Surprise PoSt. Challenge has expired.")
+	}
+
+	// 3. Verify the partialTicket values
+	if !a._verifySurprisePoStMeetsTargetReq(rt) {
+		rt.AbortStateMsg("Invalid Surprise PoSt. Tickets do not meet target.")
+	}
+
+	// verify the partialTickets themselves
+	// 4. Verify appropriate randomness
+
+	// pull from consts
+	SPC_LOOKBACK_POST := uint64(0)
+	randomnessEpoch := st.ChallengeStatus().LastChallengeEpoch()
+	randomness := rt.Randomness(randomnessEpoch, SPC_LOOKBACK_POST)
+	panic(randomness)                       // ignore circular dependency
+	var postRandomnessInput util.Randomness // sms.PreparePoStChallengeSeed(randomness, actorAddr)
```

Comment:
I don't quite understand what's going on here, what's in/complete and whether it should be straightforward in implementation.

Reply:
it should be but I'll use this to clean up randomness in the vm